Documentation for react properties (#342)
ckeditor/ckeditor4-docs: docs/guide/dev/integration/react/README.md

Please note that this property is initialised asynchronously, after mounting the component.

## Component Properties

### `config`

`Object`

Custom configuration can be passed to the editor with the `config` property of the CKEditor 4 React component. The following example shows {@link features/toolbar/README how to change the contents of the toolbar}:

```jsx
<CKEditor
	data="<p>Editor's content</p>"
	config={ {
		toolbar: [ [ 'Bold' ] ]
	} }
/>
```

{@linkapi CKEDITOR.config All configuration} options can be changed this way.

Defaults to `{}`.

### `data`

`String`

Data to be used to initially fill editor content. It's passed to {@linkapi CKEDITOR.editor#setData} method with no additional arguments.

For example usage, look at [basic usage](#basic-usage) paragraph.

Defaults to `''`.

### `name`

`String`

If provided - became editor HTML element `id` and `name` properties.

```jsx
<CKEditor
	name="myeditor"
/>
```

> Then in another place, it's possible to refer editor instance as:

Author: Rephrased :)

```js
const editor = CKEDITOR.instances.myeditor;
```

> 

Author: Added :)

> Defaults to `undefined`.

Author: Removed `undefined` default values.

Defaults to `false`.

### `style`

`Object`

Style rules set that will be applied to {@linkapi CKEDITOR.editor#container editor container} with {@linkapi CKEDITOR.dom.element#setStyles}.

```jsx
<CKEditor
	data="<p>Editor's content</p>"
	style={{
		'margin-top': '100px',
		'border': '5px solid red'
	}}
/>
```

Require [type](#type) to be `classic`.

Defaults to `undefined`.

### `type`

`'classic'` | `'inline'`

By default, the CKEditor 4 React component creates a {@link guide/dev/framed/README classic editor}. To create an {@link guide/dev/inline/README inline editor}, add the `type` property with the value of `inline` to the `<CKEditor />` tag:

```jsx
<CKEditor
	data="<p>Some initial data</p>"
	type="inline"
/>
```

You can also explicitly set the `type` property to `classic` to create the classic editor:

```jsx
<CKEditor
	data="<p>Some initial data</p>"
	type="classic"
/>
```

Every other value of the `type` property will be treated as `classic`.

For more details, look at {@link guide/dev/ckeditor_js_load/README Loading CKEditor Script article}.

Defaults to `'classic'`.

### `onBeforeLoad`

`Function`

> Callback function with single argument: `CKEDITOR` namespace. It is invoked **each time new editor instance is loaded**, but `CKEDITOR` object always refers to the same namespace.

Reviewer: Let's add a note that for modifying the `CKEDITOR` namespace it is recommended to use `onNamespaceLoaded` property instead.

Author: I added a note.

> ```jsx

Reviewer: On one hand giving the same example in both `on*` properties allows for easier comparison. However, currently we lack a good usage example of the `onBeforeLoad` property 🤔  So before the misusage, let's add a proper use case before - something basic, like logging to the console:

```jsx
<CKEditor
	data="<p>First editor.</p>"
	onBeforeLoad={ CKEDITOR => {
        console.log( 'First editor loaded!' );
	}
/>

<CKEditor
	data="<p>Second editor.</p>"
	onBeforeLoad={ CKEDITOR => {
        console.log( 'Second editor loaded!' );
	}
/>
```

Author: Added simple example and split simple and complex examples into two separate sections.

<CKEditor
	name="editorOne"
	data="<p>Editor's content</p>"
	onBeforeLoad={ CKEDITOR => {
			// Add external `placeholder` plugin which will be available for each
			// editor instance on the page.
			CKEDITOR.plugins.addExternal( 'placeholder', '/path/to/the/placeholder/plugin', 'plugin.js' );
		}
	}
/>

<CKEditor
	name="editorTwo"
	data="<p>Editor's content</p>"
	onBeforeLoad={ CKEDITOR => {

> 			// Namespace was already loaded, but this callback will be called after all.

Reviewer: ```suggestion
			// Namespace was already loaded, but this callback will be called anyway.
```

Author: Corrected.

			// Unnecessary adding same plugin to namespace.
			CKEDITOR.plugins.addExternal( 'placeholder', '/path/to/the/placeholder/plugin', 'plugin.js' );
		}
	}
/>
```

> The above example tries to load the same plugin two times to the same namespace.

Reviewer: It's more readable when the code description is given above the snippet.

Author: Moved description above the code snippet.

> Look at [onNamespaceLoaded](#onnamespaceloaded) to compare behaviours.

Author: Note added.

**Note**: To modify `CKEDITOR` namespace it is recommended to use [onNamespaceLoaded](#onnamespaceloaded) event.

Simple usage:

```jsx
<CKEditor
	data="<p>First editor.</p>"
	onBeforeLoad={ CKEDITOR => {
		console.log( 'First editor loaded!' );
	}
/>

<CKEditor
	data="<p>Second editor.</p>"
	onBeforeLoad={ CKEDITOR => {
		console.log( 'Second editor loaded!' );
	}
/>
```

An example below shows incorrect usage - it attempts to load the same plugin two times to the same namespace:

```jsx
<CKEditor
	name="editorOne"
	data="<p>Editor's content</p>"
	onBeforeLoad={ CKEDITOR => {
			// Add external `placeholder` plugin which will be available for each
			// editor instance on the page.
			CKEDITOR.plugins.addExternal( 'placeholder', '/path/to/the/placeholder/plugin', 'plugin.js' );
		}
	}
/>

<CKEditor
	name="editorTwo"
	data="<p>Editor's content</p>"
	onBeforeLoad={ CKEDITOR => {
			// Namespace was already loaded, but this callback will be called anyway.
			// Unnecessary adding the same plugin to the namespace.
			CKEDITOR.plugins.addExternal( 'placeholder', '/path/to/the/placeholder/plugin', 'plugin.js' );
		}
	}
/>

> ```

Reviewer 2: I'm not sure about giving incorrect usage example. Some people skim docs looking for code samples only so may wrongly assume it is correct usage. OTOH the code comments explains what's going on :thinking: WDYT @sculpt0r @Dumluregn?

Reviewer: Maybe we could use the warning box styling here to be sure we catch user's attention?

Author: > Some people skim docs looking for code samples only so may wrongly assume it is correct usage.

> Maybe we could use the warning box styling here to be sure we catch user's attention?

Agree :)

Reviewer 2: :+1:

Look at [onNamespaceLoaded](#onnamespaceloaded) property to compare behaviours.

### `onNamespaceLoaded`

`Function`

> Callback function with single argument: `CKEDITOR` namespace. **Invoked single time after namespace loaded, regardless of editor instances number**. This property should be used if you need to modify the `CKEDITOR` object, e.g. add an external plugin:

Reviewer 2: Maybe it will be good to mention that this event is fired before any CKEditor 4 instance is initialized? So it gives the context that modifying namespace config will affect all editor instances?

Author: I rephrase this sentence.

Please note that this property is initialised asynchronously, after mounting the component.

> ## Properties

Reviewer: ```suggestion
## Component Properties
```

Author: Auto-applied changes.

> CKEditor4 React integration supports properties:

Author: I removed an initial list of properties. Moved properties types into their descriptions.

- [config](#config) : `Object`
- [data](#data) : `''`
- [name](#name) : `String`
- [readOnly](#readonly) : `Boolean`
- [style](#style) : `Object`
- [type](#type) : `'classic'` | `'inline'`
- [onBeforeLoad](#onbeforeload) : `Function`
- [onNamespaceLoaded](#onnamespaceloaded) : `Function`

> ### config

Author: I add backticks for every property header.

> **Default**: `{}`

Author: Moved defaults to the bottom of property description. Also, add `Defaults to undefined` for `style` and `name`.

> Stores configuration settings. All available configuration options can be found in the API documentation. Refer to the {@linkapi CKEDITOR.config config} object definition. This config will be applied to editor instances separately.

Author: Auto-applied changes.

### data

**Default**: `''`

> Data will be used to initially fill editor content. It's passed to {@linkapi CKEDITOR.editor#setData} but without any other arguments.

Author: Auto-applied changes.

### name

> Provided name became editor HTML element `id` and `name` properties.

Author: Add use case provided in https://github.com/ckeditor/ckeditor4-react/issues/125.

### readOnly

**Default**: `false`

> Whenever editor should be read-only. Overrides {@linkapi CKEDITOR.config#readOnly config readOnly} before it's used to initialize editor instance.

Author: I mean hear to describe this behaviour:
https://github.com/ckeditor/ckeditor4-react/blob/4a4159dbd03bb10c206888a748bd261842e5c510/src/ckeditor.jsx#L55-L56

So no matter what config property says about `readOnly` - it's overridden, and then config is used to initialize editor.

Or you mean I described it in a not understandable manner?

> Whenever editor should be read-only. Overrides {@linkapi CKEDITOR.config#readOnly config readOnly} before it's used to initialize editor instance.

Author: It might sound not clear enough. I decided to reorganize the entire doc. I moved the already existing section about `readOnly` into property description with some adjustment.

I also moved editor `type` and `config` into properties. There were already provided samples and more proper descriptions.

### style

> Style rules set that will be applied to {@linkapi CKEDITOR.editor#container editor container} with {@linkapi CKEDITOR.dom.element#setStyles}.

Author: I removed single line breaks by joining text with previous line or split it by another empty line.

> Require [type](#type) to be different than `inline`.

Author: I use the explicit word `classic` for this line.

### type

**Default**: `'classic'`

Set editor to work in classic (iframe-based) or inline mode.

> For more details, look at {@link guide/dev/ckeditor_js_load/README Loading CKEditor Script}.

Author: I added noun `article` in link. Also format line breaks.

> ### onBeforeLoad

Author: I feel the same about `@deprecated` annotation for `onBeforeLoad`.

Added samples like described here: 
https://github.com/ckeditor/ckeditor4-docs/pull/342#discussion_r542355853

> Callback function with single argument: `CKEDITOR` namespace. **Invoked each time new editor instance is loaded**, but `CKEDITOR` always refer to the same namespace. Look at [onNamespaceLoaded](#onnamespaceloaded) to compare behaviors.

Author: Auto-applied changes.

> ### onNamespaceLoaded

Author: I added examples with adding an external plugin for two editor instances.
I used the same example in both `onBeforeLoad` and `onNamespaceLoaded`, but with different comments.

Hope that will show the difference clear enough.
